Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -355,12 +348,76 @@ void TextureMapperLayer::paint(TextureMapper& textureMapper)
     destroyFlattenedDescendantLayers();
 }
 
-#if ENABLE(WPE_PLATFORM) || PLATFORM(GTK)
-void TextureMapperLayer::collectDamage(TextureMapper& textureMapper)
+#if ENABLE(DAMAGE_TRACKING)
+void TextureMapperLayer::setDamage(const Damage& damage)
+{
+    m_damage = damage;
+}
+
+void TextureMapperLayer::collectDamage(TextureMapper& textureMapper, Damage& damage)
 {
     TextureMapperPaintOptions options(textureMapper);
     options.surface = textureMapper.currentSurface();
-    collectDamageRecursive(options);
+    collectDamageRecursive(options, damage);
+}
+
+void TextureMapperLayer::collectDamageRecursive(TextureMapperPaintOptions& options, Damage& damage)
+{
+    if (!isVisible())
+        return;
```

Comment:
I think it should be possible to return early here if the `damage` passed as argument is already invalid, because adding to it is a no-op and it would still be invalid&mdash;and we would avoid a subtree traversal. We have to be careful to reset `m_damage = Damage()` if we return early because that is currently done inside `::collectDamageSelf()`.

Maybe something like this:

```suggestion
    if (!isVisible())
        return;
        
    if (damage.isInvalid()) {
        m_damage = Damage();
        return;
    }
```

Reply:
Don't we need to clear damage in the descendants too?

Reply:
And I think it's never invalid

Reply:
Right, I double checked and that's the case because `::collectDamage()` happens to be called always with a fresh `Damage` instance in `CoordinatedGraphicsScene::paintToCurrentGLContext()`. But then there's a check that we can remove from there, I'll add another comment in the proper location.